src/dotnet/MsbuildProject.cs

```diff
@@ -166,12 +174,97 @@ public static void EnsureAllReferencesExist(List<string> references)
             }
         }
 
+        public IEnumerable<NuGetFramework> GetTargetFrameworks()
```

Comment:
I had a similar feeling about the factory methods above. Moving them into an explicit factory class would separate concerns and reduce context in each concern.